File: fhg-api/src/main/java/de/jbamberger/api/FhgApiImpl.java
package de.jbamberger.api;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.content.Context;
import android.support.annotation.NonNull;

import java.util.concurrent.atomic.AtomicBoolean;

import javax.inject.Inject;

import de.jbamberger.api.data.VPlan;
import de.jbamberger.api.data.VPlanDay;

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */

class FhgApiImpl implements FhgApi {

    private FhgEndpoint endpoint;

    @Inject
    FhgApiImpl(@NonNull Context context) {
        this.endpoint = NetModule.getEndpoint(context);
    }

    @Override
    public LiveData<ApiResponse<VPlan>> getVPlan() {
        LiveData<ApiResponse<VPlanDay>> day1 = endpoint.getVPlanFrame1();
        LiveData<ApiResponse<VPlanDay>> day2 = endpoint.getVPlanFrame2();
        AtomicBoolean isLoaded = new AtomicBoolean(false);
        final VPlan.Builder builder = new VPlan.Builder();
        final MediatorLiveData<ApiResponse<VPlan>> merger = new MediatorLiveData<>();
        merger.addSource(day1, response -> {
            merger.removeSource(day1); // there is only one value, i.e. we don't need this anymore

            if (response != null && response.isSuccessful() && response.body != null) {
                builder.addDay1(response.body);
                if (isLoaded.getAndSet(true)) {
                    merger.setValue(new ApiResponse<>(builder.build(), response));
                }
            } else {
                merger.removeSource(day2);
                if (response != null) {
                    merger.setValue(new ApiResponse<>(new Throwable(response.errorMessage)));
                } else {
                    merger.setValue(new ApiResponse<>(new Throwable("Network error")));
                }
            }
        });
        merger.addSource(day2, response -> {
            merger.removeSource(day2); // there is only one value, i.e. we don't need this anymore

            if (response != null && response.isSuccessful() && response.body != null) {
                builder.addDay2(response.body);
                if (isLoaded.getAndSet(true)) {
                    merger.setValue(new ApiResponse<>(builder.build(), response));
                }
            } else {
                merger.removeSource(day1);
                if (response != null) {
                    merger.setValue(new ApiResponse<>(new Throwable(response.errorMessage)));
                } else {
                    merger.setValue(new ApiResponse<>(new Throwable("Network error")));
                }
            }
        });

        return merger;
    }
}

File: fhg-api/src/main/java/de/jbamberger/api/FhgApiImpl.kt
package de.jbamberger.api

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MediatorLiveData
import android.content.Context
import de.jbamberger.api.data.FeedChunk
import de.jbamberger.api.data.VPlan
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */

internal class FhgApiImpl @Inject
constructor(context: Context) : FhgApi {

    private val endpoint: FhgEndpoint = NetModule.getEndpoint(context)

    override val vPlan: LiveData<ApiResponse<VPlan>>
        get() {
            val day1 = endpoint.vPlanFrame1
            val day2 = endpoint.vPlanFrame2
            val isLoaded = AtomicBoolean(false)
            val builder = VPlan.Builder()
            val merger = MediatorLiveData<ApiResponse<VPlan>>()
            merger.addSource(day1) { response ->
                merger.removeSource(day1)

                if (response != null && response.isSuccessful && response.body != null) {
                    builder.addDay1(response.body)
                    if (isLoaded.getAndSet(true)) {
                        merger.value = ApiResponse(builder.build(), response)
                    }
                } else {
                    merger.removeSource(day2)
                    if (response != null) {
                        merger.setValue(ApiResponse(Throwable(response.errorMessage)))
                    } else {
                        merger.setValue(ApiResponse(Throwable("Network error")))
                    }
                }
            }
            merger.addSource(day2) { response ->
                merger.removeSource(day2)

                if (response != null && response.isSuccessful && response.body != null) {
                    builder.addDay2(response.body)
                    if (isLoaded.getAndSet(true)) {
                        merger.value = ApiResponse(builder.build(), response)
                    }
                } else {
                    merger.removeSource(day1)
                    if (response != null) {
                        merger.setValue(ApiResponse(Throwable(response.errorMessage)))
                    } else {
                        merger.setValue(ApiResponse(Throwable("Network error")))
                    }
                }
            }

            return merger
        }

    override val feed: LiveData<ApiResponse<FeedChunk>>
        get() = endpoint.feed

}
